snuba/subscriptions/executor.py

```diff
@@ -0,0 +1,38 @@
+from __future__ import annotations
+
+from concurrent.futures import Future, ThreadPoolExecutor
+
+from snuba import settings
+from snuba.api.query import ClickhouseQueryResult, parse_and_run_query
+from snuba.datasets.dataset import Dataset
+from snuba.request import Request
+from snuba.subscriptions.consumer import Tick
+from snuba.subscriptions.data import Subscription
+from snuba.subscriptions.scheduler import ScheduledTask
+from snuba.utils.metrics.timer import Timer
+
+
+executor_pool = ThreadPoolExecutor(max_workers=settings.SUBSCRIPTIONS_MAX_CONCURRENT_QUERIES)
+
+
+class Executor:
+    """
+    Responsible for running subscription queries for a given `Dataset`. Each call to
+    `execute` should be passed the `ScheduledTask[Subscription]` to run and the `Tick`
+    we're running the query for. The `Subscription` generates the `Request` object used
+    to make the query, and the result is then returned as a `Future`.
+    """
+
+    def __init__(self, dataset: Dataset):
+        self.dataset = dataset
+
+    def execute(
+        self, task: ScheduledTask[Subscription], tick: Tick, timer: Timer
+    ) -> Future[ClickhouseQueryResult]:
+        request: Request = task.task.build_request(
+            self.dataset,
+            tick.timestamps.upper,
+            tick.offsets.upper,
+            timer,
+        )
```

Comment:
This will throw immediately, rather than binding the result of the future to an error, and will likely be uncaught by the worker/consumer.

I think we'll likely want to propagate some (maybe not all) of the errors that would be raised here to the caller instead (via the stream), since the failures here can include things like malformed queries which are expected to happen and shouldn't stop the progress of the consumer (and probably should be relayed back to the end user that their query no longer is valid.)

Here as an example from the work in progress implementation of the worker that demonstrates how I expect this to work (see the comment): https://github.com/getsentry/snuba/blob/28bf739541195b59851330a29a9663a9f2936b54/snuba/subscriptions/worker.py#L36-L48

We will at some point need to figure out how to separate expected errors (mostly bad queries, other things may fall into this category) and unexpected errors (bugs, possibly some networking problems) that should stop the consumer, but I think we can punt on that until we are actually implementing the codec that translates the `Future[ClickhouseQueryResult]` into the payload that goes on the stream.